File: C++/1026_MaximumDifferenceBetweenNodeAndAncestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxAncestorDiff(TreeNode* root) {
        if (!root) return 0;
        return helper(root);
    }
    
    int helper(TreeNode* root, int mn = 10001, int mx = 0){
        if (!root){
            return mx - mn;
        }
        mx = max(mx, root->val);
        mn = min(mn, root->val);
        return max(helper(root->left, mn, mx), helper(root->right,mn,mx));
    }
};

File: 1026_MaximumDifferenceBetweenNodeAndAncestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        return self.helper(root)
        
    def helper(self, root, mn = 10001, mx = 0):
        if not root:
            return mx - mn
        mx = max(mx, root.val)
        mn = min(mn, root.val)
        return max(self.helper(root.left, mn, mx), self.helper(root.right, mn, mx))